pvdrz/pijama_legacy
# Add some type checking tests

Tests for type checking errors for #7
Handles some (not all) cases from #20 
I want to get some feedback if this style of checking is desired.

The new test module type_check handles the successful cases and its sub-module error handles the erroneous ones.
This can be merged if the structure is ok for a test foundation.

List of handled cases:
- `true` and `false` must be `Bool`s.
- numbers must be `Int`.
- Check that binary operations are rejected if the terms doesn't have the same type.
- Check that binary operations are rejected if the terms doesn't have the same type.
- Check that arithmetic operations have type `Int` and only work with integers.
- Check that boolean operations have type `Bool` and only work with booleans.
- Check that integer comparison operations have type `Bool`  and only work with integers.
- Check that equal and not equal have type `Bool` for every base type.
- Check that arithmetic operations have type `Int` and only work with integers.
- Check that boolean operations have type `Bool` and only work with booleans.
- Check that `if x do y else z end` has type `T` if `y` and `z` have type `T` and `x` has type `Bool`.
- Check that `if x do y else z end` is rejected if `y` and `z` have different types.
- Check that `if x do y else z end` is rejected if `x` is not a boolean.
